File: TheaterSuggestions/Java/SeatsSuggestionsDomain/src/main/java/com/baasie/SeatsSuggestionsDomain/SeatCollectionExtensions.java
package com.baasie.SeatsSuggestionsDomain;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public final class SeatCollectionExtensions {

    private SeatCollectionExtensions() {
    }

    public static List<Seat> selectAvailableSeatsCompliant(List<Seat> seats, PricingCategory pricingCategory) {
        return seats.stream().filter(seat -> seat.isAvailable() && seat.matchCategory(pricingCategory)).collect(Collectors.toList());
    }

    public static List<AdjacentSeats> selectAdjacentSeats(List<Seat> seats, int partySize) {
        List<AdjacentSeats> adjacentSeatsGroups = new ArrayList<>();
        List<Seat> adjacentSeats = new ArrayList<>();

        if (partySize == 1) {
            return seats.stream().map(seat -> new AdjacentSeats(Stream.of(seat).collect(Collectors.toCollection(ArrayList::new)))).collect(Collectors.toList());
        }

        List<Seat> sortedSeatByDistanceFromCentroid = seats.stream()
                .sorted(Comparator.comparing(Seat::distanceFromCentroid))
                .collect(Collectors.toCollection(ArrayList::new));

        for (Seat candidateSeat : seats) {
            if (adjacentSeats.size() == 0) {
                adjacentSeats.add(candidateSeat);
                continue;
            }

            if (candidateSeat.isAdjacentWith(adjacentSeats)) {
                adjacentSeats.add(candidateSeat);
            } else {
                if (adjacentSeats.size() == 1) {
                    adjacentSeats = Stream.of(candidateSeat).collect(Collectors.toCollection(ArrayList::new));
                } else {
                    adjacentSeatsGroups.add(ReduceAdjacentSeats(partySize, adjacentSeats));
                    adjacentSeats = Stream.of(candidateSeat).collect(Collectors.toCollection(ArrayList::new));
                }
            }
        }

        if (adjacentSeats.size() > 1) {
            adjacentSeatsGroups.add(ReduceAdjacentSeats(partySize, adjacentSeats));
        }

        if (adjacentSeatsGroups.size() == 0) {
            adjacentSeatsGroups.add(new AdjacentSeats(adjacentSeats));
        }
        return adjacentSeatsGroups.stream().filter(a -> a.size() >= partySize).collect(Collectors.toList());
    }

    private static AdjacentSeats ReduceAdjacentSeats(int partySize, List<Seat> adjacentSeats) {

        List<Seat> orderedAdjacentSeats = adjacentSeats.stream()
                .sorted(Comparator.comparing(Seat::distanceFromCentroid))
                .collect(Collectors.toCollection(ArrayList::new));

        List<Seat> sortedSeats = orderedAdjacentSeats.stream()
                .limit(partySize)
                .sorted(Comparator.comparing(Seat::number))
                .collect(Collectors.toCollection(ArrayList::new));
        return new AdjacentSeats(sortedSeats);
    }

    public static List<AdjacentSeats> orderByMiddleOfTheRow(List<AdjacentSeats> adjacentSeats,
                                                            int rowSize) {
        Map<Integer, List<AdjacentSeats>> sortedAdjacentSeatsGroups = new TreeMap<>();

        for (AdjacentSeats adjacentSeat : adjacentSeats) {
            int distance = adjacentSeat.computeDistanceFromRowCentroid();

            if (!sortedAdjacentSeatsGroups.containsKey(distance)) {
                sortedAdjacentSeatsGroups.put(distance, new ArrayList<>());
            }

            sortedAdjacentSeatsGroups.get(distance).add(adjacentSeat);
        }

        return sortedAdjacentSeatsGroups.values().stream().flatMap(Collection::stream).collect(Collectors.toList());
    }

    public static int centroidIndex(int rowSize) {
        return Math.abs(rowSize / 2);
    }

    public static int computeDistanceFromCentroid(int seatLocation, int rowSize) {
        return Math.abs(seatLocation - centroidIndex(rowSize));
    }

    public static boolean isCentroid(int seatLocation, int rowSize) {
        int centroidIndex = centroidIndex(rowSize);

        return seatLocation == centroidIndex || seatLocation == centroidIndex + 1;
    }

    public static boolean isOdd(int rowSize) {
        return rowSize % 2 != 0;
    }
}

File: TheaterSuggestions/CSharp/SeatsSuggestions.Domain/SeatCollectionExtensions.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

namespace SeatsSuggestions.Domain
{
    public static class SeatCollectionExtensions
    {
        public static IEnumerable<Seat> SelectAvailableSeatsCompliant(this IEnumerable<Seat> seats,
            PricingCategory pricingCategory)
        {
            return seats.Where(s => s.IsAvailable() && s.MatchCategory(pricingCategory));
        }

         public static IEnumerable<AdjacentSeats> SelectAdjacentSeats(this IEnumerable<Seat> seats,
            int partySize)
        {
            var adjacentSeatsGroups = new List<AdjacentSeats>();
            var adjacentSeats = new List<Seat>();
            var currentSeats = seats.ToList();

            if (partySize == 1) return currentSeats.Select(s => new AdjacentSeats(new List<Seat> {s}));

            foreach (var candidateSeat in currentSeats)
            {
                if (!adjacentSeats.Any())
                {
                    adjacentSeats.Add(candidateSeat);
                    continue;
                }

                adjacentSeats = adjacentSeats.OrderBy(s => s.Number).ToList();

                if (candidateSeat.IsAdjacentWith(adjacentSeats))
                {
                    adjacentSeats.Add(candidateSeat);

                    if (NoMoreSeats(adjacentSeats, currentSeats))
                    {
                        adjacentSeatsGroups.Add(new AdjacentSeats(ReduceAdjacentSeats(partySize, adjacentSeats)));
                    }
                }
                else
                {
                    if (adjacentSeats.Any())
                    {
                        adjacentSeatsGroups.Add(ReduceAdjacentSeats(partySize, adjacentSeats));
                    }

                    adjacentSeats = new List<Seat> {candidateSeat};
                }
            }

            return adjacentSeatsGroups.Where(adjacent => adjacent.Count() >= partySize);
        }

        private static AdjacentSeats ReduceAdjacentSeats(int partySize, List<Seat> adjacentSeats)
        {
            var orderedEnumerable = adjacentSeats.OrderBy(s => s.DistanceFromRowCentroid).ToList();

            return new AdjacentSeats(orderedEnumerable.Take(partySize).OrderBy(s => s.Number));
        }

        private static bool NoMoreSeats(ICollection adjacentSeats, ICollection currentSeats)
        {
            return adjacentSeats.Count == currentSeats.Count;
        }

        private static bool DoesNotExceedPartyRequestedAndCandidateSeatIsAdjacent(Seat candidateSeat,
            List<Seat> adjacentSeats,
            int partySize)
        {
            return candidateSeat.IsAdjacentWith(adjacentSeats) && adjacentSeats.Count < partySize;
        }

        public static IEnumerable<AdjacentSeats> OrderByMiddleOfTheRow(this IEnumerable<AdjacentSeats> adjacentSeats,
            int rowSize)
        {
            var sortedAdjacentSeatsGroups = new SortedList<int, List<AdjacentSeats>>();

            foreach (var adjacentSeat in adjacentSeats)
            {
                var distanceFromRowCentroid = adjacentSeat.ComputeDistanceFromRowCentroid(rowSize);

                if (!sortedAdjacentSeatsGroups.ContainsKey(distanceFromRowCentroid))
                    sortedAdjacentSeatsGroups.Add(distanceFromRowCentroid, new List<AdjacentSeats>());

                sortedAdjacentSeatsGroups[distanceFromRowCentroid].Add(adjacentSeat);
            }

            return sortedAdjacentSeatsGroups.Values.SelectMany(_ => _);
        }

        internal static int CentroidIndex(this int rowSize)
        {
            return (int)Math.Ceiling((decimal)rowSize / 2);
        }

        internal static int ComputeDistanceFromCentroid(this uint seatLocation, int rowSize)
        {
            return (int)Math.Abs(seatLocation - rowSize.CentroidIndex());
        }

        internal static bool IsCentroid(this uint seatLocation, int rowSize)
        {
            var centroidIndex = rowSize.CentroidIndex();
            return seatLocation == centroidIndex;
        }

        internal static bool IsOdd(this int rowSize)
        {
            return rowSize % 2 != 0;
        }
    }
}